src/repr/src/relation.rs

```diff
@@ -657,13 +819,285 @@ impl fmt::Display for NotNullViolation {
+        indices.sort_unstable();
+        if !self.keys.contains(&indices) {
+            self.keys.push(indices);
+        }
+        self
+    }
+
+    /// Removes all previously inserted keys.
+    pub fn without_keys(mut self) -> RelationDescBuilder {
+        self.keys.clear();
+        assert_eq!(self.keys.len(), 0);
+        self
+    }
+
+    /// Concatenates a [`RelationDescBuilder`] onto the end of this [`RelationDescBuilder`].
+    pub fn concat(mut self, other: Self) -> Self {
+        let self_len = self.columns.len();
+
+        self.columns.extend(other.columns);
+        for k in other.keys {
+            let k = k.into_iter().map(|idx| idx + self_len).collect();
+            self = self.with_key(k);
+        }
+
+        self
+    }
+
+    /// Finish the builder, returning a [`RelationDesc`].
+    pub fn finish(self) -> RelationDesc {
+        let mut desc = RelationDesc::from_names_and_types(self.columns);
+        desc.typ = desc.typ.with_keys(self.keys);
+        desc
+    }
+}
+
+/// Describes a [`RelationDesc`] at a specific version of a [`VersionedRelationDesc`].
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+pub enum RelationVersionSelector {
+    Specific(RelationVersion),
+    Latest,
+}
+
+impl RelationVersionSelector {
+    pub fn specific(version: u64) -> Self {
+        RelationVersionSelector::Specific(RelationVersion(version))
+    }
+}
+
+/// A wrapper around [`RelationDesc`] that provides an interface for adding
+/// columns and generating new versions.
+///
+/// TODO(parkmycar): Using an immutable data structure for RelationDesc would
+/// be great.
+#[derive(Debug, Clone, Serialize)]
+pub struct VersionedRelationDesc {
+    inner: RelationDesc,
+}
+
+impl VersionedRelationDesc {
+    pub fn new(inner: RelationDesc) -> Self {
+        VersionedRelationDesc { inner }
+    }
+
+    /// Adds a new column to this [`RelationDesc`], creating a new version of the [`RelationDesc`].
+    ///
+    /// Note: For building a [`RelationDesc`] see [`RelationDescBuilder::with_column`].
+    #[must_use]
+    pub fn add_column<N, T>(&mut self, name: N, typ: T) -> RelationVersion
+    where
+        N: Into<ColumnName>,
+        T: Into<ColumnType>,
+    {
+        let latest_version = self.latest_version();
+        let new_version = latest_version.bump();
+
+        let next_idx = self.inner.metadata.len();
+        let col_meta = ColumnMetadata {
+            name: name.into(),
+            typ_idx: next_idx,
+            added: new_version,
+            dropped: None,
+        };
+
+        self.inner.typ.column_types.push(typ.into());
+        let prev = self.inner.metadata.insert(ColumnIndex(next_idx), col_meta);
+
+        assert_none!(prev, "column index overlap!");
+        self.validate();
+
+        new_version
+    }
+
+    /// Drops the column `name` from this [`RelationDesc`]. If there are multiple columns with
+    /// `name` drops the left-most one.
```

Comment:
> If there are multiple columns with
>    /// `name` drops the left-most one.

Is it possible to have multiple columns with the same name?

Reply:
It is something we want to support, e.g.
```
CREATE TABLE t1 (a int);
ALTER TABLE t1 DROP COLUMN a;
ALTER TABLE t1 ADD COLUMN a text;
```
We would keep the original `a int` as `ColumnIndex(0)` and have `a text` as `ColumnIndex(1)`

Reply:
In that case if I tried to do something like

```
CREATE TABLE t1 (a int);
ALTER TABLE t1 DROP COLUMN a;
ALTER TABLE t1 ADD COLUMN a text;
ALTER TABLE ts DROP COLUMN a;
```

then the second drop would panic on line 988?

Maybe the description should read something like

> If there are multiple columns with  `name` drops the left-most non-dropped one.

and line 984 should be something like `.find(|meta| meta.name == name && meta.dropped.is_none())`